File: 1036-rotting-oranges/rotting-oranges.cpp
class Solution {
public:
    vector<vector<int>>dir={{0,1},{1,0},{0,-1},{-1,0}};
    int bfs(queue<vector<int>>&q,vector<vector<int>>&g,vector<vector<int>>&vis){
        int cnt=0;
        while(!q.empty()){
            auto top=q.front();
            int cr=top[0];
            int cc=top[1];
            int dist=top[2];
            cnt=max(dist,cnt);
            q.pop();
            for(auto x:dir){
                if(x[0]+cr<g.size()&& x[1]+cc<g[0].size() && !vis[cr+x[0]][cc+x[1]] && g[cr+x[0]][cc+x[1]]==1){
                    vis[cr+x[0]][cc+x[1]]=1;
                    q.push({cr+x[0],cc+x[1],dist+1});
                    
                }
            }
        }
        return cnt;
    }
    int orangesRotting(vector<vector<int>>& g) {
        int m=g.size();
        int n=g[0].size();
        vector<vector<int>>vis(m,vector<int>(n,0));
        queue<vector<int>>q;
        for(int i=0;i<m;i++){
            for(int j=0;j<n;j++){
                if(g[i][j]==2){
                    q.push({i,j,0});
                    vis[i][j]=1;
                }
            }
        }
        int maxi=bfs(q,g,vis);
        for(int i=0;i<m;i++){
            for(int j=0;j<n;j++){
                if(g[i][j]==1 && !vis[i][j]){
                    return -1;
                }
            }
        }
        return maxi;
        
    }
};

File: 1036-rotting-oranges/rotting-oranges.java
class triplet{
    int first;
    int second;
    int third;
    triplet(int first,int second,int third){
        this.first=first;
        this.second=second;
        this.third=third;
    }
}
class Solution {
    int [][] dirs ={{-1,0},{1,0},{0,1},{0,-1}};
    public int orangesRotting(int[][] grid) {
        int m=grid.length;
        int n=grid[0].length;
        Queue<triplet> q = new LinkedList<>();
        int [][] vis = new int[m][n];
        int max_time=0;
        for(int i=0;i<m;i++){
            for(int j=0;j<n;j++){
                if(grid[i][j]==2){
                    q.add(new triplet(i,j,0));
                    vis[i][j]=1;
                }
            }
        }
        while(!q.isEmpty()){
            triplet t = q.poll();
            int i = t.first;
            int j = t.second;
            int time= t.third;
            max_time=Math.max(time,max_time);
            for(int d=0;d<4;d++){
                int [] dir = dirs[d];
                int nr=i+dir[0];
                int nc=j+dir[1];
                if(nr<m && nr>=0 && nc>=0 && nc<n && vis[nr][nc]==0 && grid[nr][nc]==1){
                    vis[nr][nc]=1;
                    q.add(new triplet(nr,nc,time+1));
                }
            }
        }
        for(int i=0;i<m;i++){
            for(int j=0;j<n;j++){
                if(grid[i][j]==1 && vis[i][j]==0)return -1;
            }
        }
        return max_time;

    }
}